internal/uvm/vpmem.go

```diff
@@ -2,60 +2,301 @@ package uvm
+	return vpm.refCount
+}
+
+func (vpm *vpmemMapping) AddRef() uint32 {
+	vpm.refCount++
+	return vpm.refCount
+}
+
+func (vpm *vpmemMapping) RemoveRef() uint32 {
+	vpm.refCount--
+	return vpm.refCount
+}
+
+func (vpm *vpmemMapping) HostPath() string {
+	return vpm.hostPath
+}
+
+func (vpm *vpmemMapping) MountPath() string {
+	if vpm.uvmPath != "" {
+		return vpm.uvmPath
+	}
+	return fmt.Sprintf(lcowVPMemLayerMountFmt, vpm.deviceNumber, vpm.deviceOffset, vpm.deviceSize)
+}
+
+func (vpm *vpmemMapping) ResourcePath() string {
+	return fmt.Sprintf(vPMemDeviceResourceFormat, vpm.deviceNumber, vpm.deviceOffset)
+}
+
+func (vpm *vpmemMapping) DeviceNumber() uint32 {
+	return vpm.deviceNumber
+}
+
+func (vpm *vpmemMapping) DeviceSize() uint64 {
+	return vpm.deviceSize
+}
+
+func (vpm *vpmemMapping) DeviceOffset() uint64 {
+	return vpm.deviceOffset
+}
+
+func (vpm *vpmemMapping) HostSettings() interface{} {
+	return hcsschema.VirtualPMemMapping{
+		HostPath:    vpm.hostPath,
+		ImageFormat: "Vhd1",
+	}
+}
+
+func (vpm *vpmemMapping) GuestSettings() interface{} {
+	return guestrequest.LCOWMappedVPMemDevice{
+		DeviceNumber: vpm.deviceNumber,
+		MountPath:    vpm.MountPath(),
+		MappingInfo: guestrequest.LCOWMappedLayer{
+			DeviceOffsetInBytes: vpm.deviceOffset,
+			DeviceSizeInBytes:   vpm.deviceSize,
+		},
+	}
+}
+
+var _ mappedDevice = &vpmemMapping{}
+
+// newVPMemDevice creates a new vpmemDevice instance and allocates a single 4GB memory slot
+func newVPMemDevice() *vpmemDevice {
+	dev := &vpmemDevice{
+		memoryAllocator: newDefaultMemoryAllocator(),
+		maxSize:         DefaultVPMemSizeBytes,
+		mappings:        make(map[string]mappedDevice),
+	}
+	return dev
+}
+
+func newVPMemMapping(devNumber uint32, devSize, devOffset uint64, hostPath string) *vpmemMapping {
+	return &vpmemMapping{
+		deviceNumber: devNumber,
+		deviceSize:   devSize,
+		deviceOffset: devOffset,
+		hostPath:     hostPath,
+		refCount:     1,
+	}
+}
+
+// mapDevice increases reference count if `device` exists or adds a new mapping
+func (vpd *vpmemDevice) mapDevice(ctx context.Context, device mappedDevice) (err error) {
+	if dev, ok := vpd.mappings[device.HostPath()]; ok {
+		dev.AddRef()
+		return nil
+	}
+
+	// Make sure that the requested offset is matching the first available one
+	var offset uint64
+	memCls := getMemoryClassType(device.DeviceSize())
+	offset, err = vpd.findFreeOffset(memCls)
+	if err != nil {
+		return err
+	}
+	if offset != device.DeviceOffset() {
+		return errors.Errorf("device offset mismatch: available=%d, requested=%d", offset, device.DeviceOffset())
+	}
+
+	// Allocate memory slot and make sure that the expected allocation happened
+	var slot *memSlot
+	slot, err = vpd.allocate(memCls)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			vpd.release(memCls, slot.offset)
+		}
+	}()
+
+	if slot.offset != device.DeviceOffset() {
+		return errors.Errorf("unexpected slot allocation: expected=%d, actual=%d", offset, slot.offset)
+	}
```

Comment:
Is this possible?

Reply:
not if the caller is playing by the rules lol, but we could potentially get a `device` that will have some random offset in it